### PR TITLE
Compression function inlining/flattening argument/var names colliding with argument values inner scopes

### LINKED ISSUE #292: Compression function inlining / flattening does not always produce working code
**Bug report or Feature request?**
Bug

**Version (complete output of `terser -V`)**
terser 3.16.1

**Complete CLI command or `minify()` options used**
`terser test.js -c --toplevel`

**`terser` input**
`test.js` contains the following code:
```
var g1 = ['a'];

function problem(arg1) {
  return g1.indexOf(arg1);
}

function unused(arg1) {
  return problem(arg1);
}

function B(problem) {
  return g1[problem];
}

function A(arg1) {
  return problem(arg1);
}

function main(arg1) {
  return B(A(arg1));
}

console.log(main('a'));
```

This code has no useful functionality, it's just a very stripped down piece of code that still produces the issue I'm having with my actual code.

**`terser` output or error**
Terser output:
`var g1=["a"];function problem(arg1){return g1.indexOf(arg1)}console.log(function(arg1){return problem=function(arg1){return problem(arg1)}(arg1),g1[problem];var problem}("a"));`

Running this code will lead to the following error:
`TypeError: problem is not a function`

Compression does not seem to take into account an argument overriding a toplevel function when inlining / flattening.

**Expected result**
This code should do the same as:
`console.log("a");`

What the exact proper output should be, is hard to say. If you can already change names during compression, this output could work:
`var g1=["a"];function problem(arg1){return g1.indexOf(arg1)}console.log(function(arg1){return noproblem=function(arg1){return problem(arg1)}(arg1),g1[noproblem];var noproblem}("a"));`

But if renaming is not an option, then probably skip the inlining like so:
`var g1=["a"];function problem(arg1){return g1.indexOf(arg1)}console.log(function(arg1){return function(problem){return g1[problem];}(function(arg1){return problem(arg1)}(arg1));}("a"));`

### PR DESCRIPTION
**Pull request by j-verse**
This is the first time I've tried to work on the inner workings of uglify-js / terser. I have some compiler knowledge that helped with understanding the AST and how the compressor works. The solution I came up with to fix #292 does work for my test cases and for the original problem I had, however I can't guarantee it will work for all edge cases. There might also be better ways to solve the problem, but this was the best I could come up with for now. One good thing about how I solved the issue, it only affects the decision about whether to inline / flatten a function, overall working of compression should not be affected other than the cases meeting the conditions where flattening should not happen.

Comments are very welcome :)

### CODE REVIEW

File: lib/compress/index.js [hunk 1]
```diff
@@ -5177,6 +5177,49 @@ merge(Compressor.prototype, {
             return true;
         }
 
+        function can_inject_args_values() {
+            var arg_fns_outer_refs = Object.create(null);
```
[the remark this answers is not in the document]
Comment (j-verse) [reply lib/compress/index.js]: Agreed, replaced the object map with a Set.

File: lib/compress/index.js [hunk 2]
```diff
@@ -5177,6 +5177,49 @@ merge(Compressor.prototype, {
...
+            var value_walker = new TreeWalker(function(node) {
+                if (node instanceof AST_Scope) {
+                    var var_list = [];
```
[the remark this answers is not in the document]
Comment (j-verse) [reply lib/compress/index.js]: True, now I just iterate over the set.

File: lib/compress/index.js [hunk 3]
```diff
@@ -5177,6 +5177,49 @@ merge(Compressor.prototype, {
...
+                    var var_outer_ref_map = Object.create(null);
+                    node.enclosed.forEach(function(def) {
+                        var_list.push(def.name);
+                        var_outer_ref_map[def.name] = true;
+                    });
+                    node.variables.each(function(def, name) {
+                        var_list.push(name);
+                        var_outer_ref_map[name] = false;
+                    });
+                    var_list.forEach(function(name) {
```
[the remark this answers is not in the document]
Comment (j-verse) [reply lib/compress/index.js]: I'm iterating over the set now, however, if I've read the contributing guidelines correctly, we can't use the fat arrow right? Since that is ES6 and the code is supposed to be ES5?

[the remark this answers is not in the document]
Comment (L2jLiga) [reply lib/compress/index.js]: Since recent commit we dropped out node 4.x support and now can use such things like Set, Map, arrow functions and other stuff supported by node 6.x and upper: https://node.green

[the remark this answers is not in the document]
Comment (j-verse) [reply lib/compress/index.js]: Should I replace all lambda's with arrow functions then? Or keep it consistent with the rest of the compress lib?

[the remark this answers is not in the document]
Comment (L2jLiga) [reply lib/compress/index.js]: I think it would be better to keep it consistent

**Comment by j-verse** [comment]
Added a check to skip checking argument names and variable names when none of the argument values refer to outer scopes.

File: lib/compress/index.js [hunk 4]
```diff
@@ -5177,6 +5177,49 @@ merge(Compressor.prototype, {
             return true;
         }
 
+        function can_inject_args_values() {
+            var arg_vals_outer_refs = new Set();
+            var value_walker = new TreeWalker(function(node) {
+                if (node instanceof AST_Scope) {
```
Comment (fabiosantoscode) [inline lib/compress/index.js]: This will not find block scopes. Block scopes are the `block_scope` property of each block.

Comment (fabiosantoscode) [reply lib/compress/index.js]: I think that sucks, but it's because Terser was built from the ground up for ES5, with the idea that a scope is always a function. This is of course not true anymore :'(

Comment (j-verse) [reply lib/compress/index.js]: I must say, I was slightly confused about how block scopes were integrated in the compressor. However I did not check for it, because an argument has to be an expression. It can only really have an inner scope either when a lambda function is passed as an argument or some expression involving an IIFE, right? I could not think of a way it would directly contain a block scope. It can of course contain a block scope inside the lambda or IIFE, but I assumed all symbols refering to outer scopes would then be mentioned in the lambda/IIFE's enclosed property.

Please correct me if I'm wrong.

Comment (j-verse) [reply lib/compress/index.js]: Anyone? It would be nice if we can conclude if my thinking is correct. If it isn't I'll try to fix it, if it is this bug is solved.

Comment (L2jLiga) [reply lib/compress/index.js]: @j-verse from my end I think that this PR is good but I'm not sure that I actually get what did @fabiosantoscode mean.

My example input:

```js
var g = ["a"];

function problem(arg) {
    return g.indexOf(arg);
}

function unused(arg) {
    return problem(arg);
}

try {
    function a(arg) {
        return problem(arg);
    }

    function b(problem) {
        return g[problem];
    }

    function c(arg) {
        return b(a(arg));
    }

    block: {
        let g = null;
        let c = Math.round(Math.random() + 1);
        switch (c) {
            case 1: {
                let b = a("a");
                console.log(b);
                break;
            }
            case 2: {
                let c = b(a("a"));
                console.log(c);
                break;
            }
            default:
                console.log("c");
        }
    }

} catch (e) {
    console.log("Error!");
}
```

And Terser output:
```js
var g = ["a"];

function problem(arg) {
    return g.indexOf(arg);
}

try {
    function a(arg) {
        return problem(arg);
    }

    function b(problem) {
        return g[problem];
    }

    switch (Math.round(Math.random() + 1)) {
        case 1: {
            let b = a("a");
            console.log(b);
            break;
        }
        case 2: {
            let c = b(a("a"));
            console.log(c);
            break;
        }
        default:
            console.log("c");
    }
} catch (e) {
    console.log("Error!");
}
```

**Comment by fabiosantoscode** [comment]
@j-verse I apologize for taking a long time to get back to this PR. I understand why your code doesn't need to think of block scopes. I think this is a valuable addition to Terser and should go in. Merging!
